### ISSUE TITLE
HorizontalStackLayout does not limit maximum width of child label to it's width

### ISSUE THREAD

**Issue by philipag** [dotnet/maui#10046]
### Description

As shown in the attached sample, the label text does not wrap and the label extends beyond the blue bounds of the parent HorizontalStackLayout.  If I change HorizontalStackLayout  to any other layout control (e.g. VerticalStackLayout) then the label wraps as expected.  The expected behavior is for the label to have the same width as the blue parent  HorizontalStackLayout. 

![image](https://user-images.githubusercontent.com/6193288/189545539-77727fc5-fb7d-4989-a4a1-821fd4b0c355.png)
[layoutbug2.zip](https://github.com/dotnet/maui/files/9543370/layoutbug2.zip)

### Steps to Reproduce

1. Deploy included sample app to Android emulator (Pixel_5)
2. Label extends to the right of the parent HorizontalStackLayout but should not.

### Link to public reproduction project repository

none

### Version with bug

6.0.486 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, I was *not* able test on other platforms

### Affected platform versions

Android 12

### Did you find any workaround?

Setting Maximum width of label will cause it to wrap, but this should not be necessary.

**Comment by philipag** [comment]
Similarly if I use a Grid with column "*" instead of HorizontalStackLayout, the Grid also becomes larger than its parent which causes the label to not wrap again.  Note this only happens when setting Grid.HorizontalOptions="Center" like so:

`	<Grid BackgroundColor="Blue" ColumnDefinitions="*" VerticalOptions="CenterAndExpand" HorizontalOptions="Center">
		<Label Grid.Column="0" VerticalOptions="Center" Text="this is a label which should wrap" FontSize="40" BackgroundColor="#55ff0000"
				   LineBreakMode="WordWrap"
			   HorizontalTextAlignment="Center"  VerticalTextAlignment="Center"
					/>
	</Grid>`

Setting Grid.HorizontalOptions="Fill" or something other than Center causes the label to wrap normally.  Perhaps I don't understand the Center layout flag but it seems that this is a bug related to this issue.

**Comment by hartez** [comment]
> Label extends to the right of the parent HorizontalStackLayout but should not.

Are you trying to wrap the text of the Label, or clip the Label at the edge?

**Comment by philipag** [comment]
@hartez  LineBreakMode="WordWrap" so the label should wrap.  And it does if e.g. one changes the HorizontalStackLayout to a VerticalStackLayout.

**Comment by petrakritzinger** [comment]
Having the same issue, no workaround available. Breaking most of the layouts in our app. Please let us know when there is a workaround or a fix.

**Comment by philipag** [comment]
I'm afraid that this  might be by design.  According to [https://github.com/dotnet/maui/wiki/Xamarin.Forms----MAUI.Controls-Layout-Differences](url) the MAUI stack layout allow their children to extend beyond their size (in the stacking direction).  I had to switch to using Grid for these cases.

**Comment by petrakritzinger** [comment]
I don't agree that this is by design. "Allowing" and "forcing" their children to extend beyond their size is not the same thing. Also, the text does wrap with a vertical stacklayout, just not with horizontal.

**Comment by philipag** [comment]
I don't like this behavior either and it caused quite a bit of rework when migrating from XF.  Maybe someone who understands the underlying details can look at this and shed some light for us?

**Comment by ArturWyszomirski** [comment]
I've got the same issue with label inside a ScrollView (that spans over all columns) inside a Grid. It worked diffrent in .NET 6 - the text in label was wrapped by default. 

Ok, now I've noticed that the problem is when any column width is set to Auto. When it's set to * or some numeric value the text is wrapped by default.

**Comment by jinxinjuan** [comment]
Verified this issue with Visual Studio Enterprise 17.7.0 Preview 1.0. Can repro on Android platform (API-33) with sample project. [layoutbug2.zip](https://github.com/dotnet/maui/files/9543370/layoutbug2.zip)

**Comment by navneettavant** [comment]
Same issue still observered in .Net 7 & MAUI 7 version. 
Please let me know is it fixed or not. 
Same behaviour exist in IOS application. 
I have a below structure 
Listview->Stacklayout-> Label with LineBreakMode="WordWrap" and maxline=2 but still text is not wrapped in label it goes out of the screen. 
Without listview it works fine but inside listview it creates an issue.

**Comment by albyrock87** [comment]
I have the strong feeling this would be fixed by #20133

**Comment by MauiUIui** [comment]
Having a label inside a `Grid`, `VerticalStackLayout `or whatever doesn't fix the problem that I have.
It seems as if it has something to do with the fonts used (or FontAttributes or TextDecorations no matter what container, see explanation below). 
See I have a label (Android) which uses an exotic font. I have `LineBreakMode="WordWrap" MaxLines="3"`
Now my actual text length depends on the language used, as the app is (statically) localized and it is never actually 3 lines long (either 1 or 2). 
For clarification the text is 2-3 words long depending on the language and I have `FontSize="38"`. So that's why this is critical, as it's most likely only one word that would have to be sent to the next line.
Most of the time everything works fine. The line breaks when needed and the label is fully shown. However sometimes (really, this is very text depended) the line apparently does break but only the first line of the label is shown (or let's say the last line is not showing) so mostly the last word is not shown at all. Now if I change the language to another one where the text is also 2 lines long, restart the app, the same label is fully shown.
My understanding is that there are at least two areas where the needed width for the label is being calculated. However two (or more) different methods of calculation are being used. Could it be, that one is taking the font into consideration and one is not? Or maybe just the `TextDecorations` or `FontAttributes` are not being considered in one method (the one the label container uses) while the label considers them? The later is very likely, since I already noticed a problem when I have `FontAttributes="Italic"` on other occasions where a part of the last letter of the text is cut off in some standard fonts (most fonts tbh) and it really looks like as if the container is giving the label just a little bit too few of space. By the way, on the label from the example above I also have `FontAttributes="Bold"`
@PureWeen Is that enough to accelerate working on this bug as it does break my layout (and my heart). I very much think that it's one little thing that'd fix many similar issues.

**Comment by mkshiblu** [comment]
Any update on this? Still present in .NET 9

**Comment by wbv-profile** [comment]
Still reproduces on `main` @ `bc3095fa`.

### Mechanism

`HorizontalStackLayoutManager.Measure` offers every child an **unbounded** width, regardless of what the stack itself was given:

https://github.com/dotnet/maui/blob/main/src/Core/src/Layouts/HorizontalStackLayoutManager.cs#L30

```csharp
var measure = child.Measure(double.PositiveInfinity, heightConstraint - padding.VerticalThickness);
```

The child's desired width is then summed into `measuredWidth`, and `ResolveConstraints` clamps **the stack's own reported size** to the incoming constraint — but nothing clamps the child. So the stack tells its parent it fits while arranging the child outside itself.

### Repro

Against `src/Core/tests/UnitTests` (no device needed):

```csharp
var child = LayoutTestHelpers.CreateTestView(new Size(1000, 100));
var stack = CreateTestLayout(new List<IView> { child });

var manager = new HorizontalStackLayoutManager(stack);
var measured = manager.Measure(300, 100);          // stack constrained to 300
manager.ArrangeChildren(new Rect(0, 0, 300, 100));
```

Observed:

| | value |
|---|---|
| stack constraint | 300 |
| width offered to child | **∞** |
| stack's measured width | 300 |
| child's arrange rect | **1000** x 100 |

The child is laid out 700pt beyond the stack's bounds.

`VerticalStackLayoutManager.cs#L31` does the same on the height axis.

### Related

#23735 is the same defect in `GridLayoutManager`: an `Auto` column sets `cell.MeasureWidth = double.PositiveInfinity` (`GridLayoutManager.cs#L1052`) and likewise never clamps the child back to the container's constraint. Same repro shape, same result — child offered ∞, arranged at 1000 inside a 300pt grid. Worth fixing together, since a fix to one that doesn't address the other will leave the nested `Grid`-inside-`StackLayout` cases still broken.

**Comment by hartez** [comment]
This is by design. HorizontalStackLayout and VerticalStackLayout do not constrain their children in the horizontal/vertical dimensions. This issue should be closed.
